### PR TITLE
TO_JSON for Mojo::Collection

### PR DESCRIPTION
**Pull request by wttw**
Adds Mojo::Collection::TO_JSON

The name is slightly misleading as it doesn't convert to JSON, rather it returns an array reference in the same way as to_array. This lets Mojo::JSON and friends convert a Mojo::Collection to a JSON array.

Amongst other things it reduces noise when returning Mojo::Pg result sets as JSON:

    $c->render(json => {foo => $results->hashes});

vs

    $c->render(json => {foo => $results->hashes->to_array});

### CODE REVIEW

File: lib/Mojo/Collection.pm [hunk 1]
```diff
@@ -90,6 +90,8 @@ sub tap { shift->Mojo::Base::tap(@_) }
 
 sub to_array { [@{shift()}] }
 
+sub TO_JSON { [@{shift()}] }
```
Comment (kraih) [inline lib/Mojo/Collection.pm]: Upper case methods should be at the top.

File: lib/Mojo/Collection.pm [hunk 2]
```diff
@@ -322,6 +324,16 @@ Alias for L<Mojo::Base/"tap">.
 
 Turn collection into array reference.
 
+=head2 TO_JSON
+
+  my $array = $collection->TO_JSON;
+
+Turn collection into array reference allowing conversion to JSON by
+L<Mojo::JSON>.
```
Comment (kraih) [inline lib/Mojo/Collection.pm]: It is merely an alias for `to_array`, that's what the description should say (and link to).

File: lib/Mojo/Collection.pm [hunk 3]
```diff
@@ -322,6 +324,16 @@ Alias for L<Mojo::Base/"tap">.
...
+
+  # {"baz":["foo","bar"]}
+  $c->render(json => { baz => Mojo::Collection->new('foo', 'bar') });
```
Comment (kraih) [inline lib/Mojo/Collection.pm]: This example is out of context.

**Comment by kraih** [comment]
:-1: from me i'm afraid. Like i said on IRC before, we've had a lot of bad luck with this kind of magic in `Mojo::Collection` before, which makes me very sceptical of any new additions. There is also an inconsistency between `Mojo::ByteStream` and `Mojo::Collection` now, since both would behave differently when passed to JSON encoders like `Cpanel::JSON::XS`.

**Comment by kraih** [comment]
Just in case some of you don't remember those magical mistakes... we've had `Mojo::Collection` return true or false in boolean context depending on the number of elements, stringify all elements joined with newlines, and allowed arbitrary methods to be called on all elements with `AUTOLOAD`. And every single one of those turned out to be a huge problem.

**Comment by Grinnz** [comment]
I would be +1 if accompanied by a matching method for Mojo::ByteStream, and with the docs fixed up. I don't believe this is a problem like the other Mojo::Collection issues, as this only affects JSON encoding unlike overloading and AUTOLOAD, and treats the object as an arrayref and not an array.

**Comment by tempire** [comment]
It doesn't seem so bad to me, given that the "magic" is only what Mojo::JSON already does. I'm :+1:, but open to being convinced otherwise.

File: lib/Mojo/Collection.pm [hunk 4]
```diff
@@ -322,6 +324,12 @@ Alias for L<Mojo::Base/"tap">.
 
 Turn collection into array reference.
 
+=head2 TO_JSON
```
Comment (kraih) [inline lib/Mojo/Collection.pm]: This section needs to be moved to the top too.

**Comment by marcusramberg** [comment]
I'm :-1: for now because of the inconsistency it adds, I'm not as worried about the "magic"/convention of TO_JSON tho. I might be inclined to change my vote if Mojo::ByteStream got the TO_JSON method too.

**Comment by jhthorsen** [comment]
:+1: from me. Same as @Grinnz's comments. I would actually stretch it to considering adding TO_JSON() to all the modules that does to_array or to_string, to make other mojo-objects consistent with Cpanel::JSON::XS.

**Comment by jberger** [comment]
Like @jhthorsen I think that many of the classes that implement to_string (and to_array, but there's only one of those) ought to have TO_JSON, except that I don't think they all should. Here is the list of those:

```bash
$ ack -l 'sub to_string'
lib/Mojo/ByteStream.pm
lib/Mojo/Cookie/Request.pm
lib/Mojo/Cookie/Response.pm
lib/Mojo/Cookie.pm
lib/Mojo/Date.pm
lib/Mojo/DOM.pm
lib/Mojo/Exception.pm
lib/Mojo/Headers.pm
lib/Mojo/Home.pm
lib/Mojo/Message.pm
lib/Mojo/Parameters.pm
lib/Mojo/Path.pm
lib/Mojo/URL.pm
lib/Mojolicious/Routes/Route.pm
```

While I can see the need to serialize several of these, I don't think we need to really concern ourselves with all of them (would someone really want to serialize a route?). That said, I'm now choosing between consistency and practicality, which I know will displease @kraih. I still remain neutral I suppose, though with some amount of sadness.

**Comment by kraih** [comment]
I have to strongly disagree with @jhthorsen, adding a `TO_JSON` method to classes like `Mojo::Message` would be a huge mistake.

**Comment by jhthorsen** [comment]
Sorry! Let me rephrase. I meant to make it consistent between Mojo::JSON and Cpanel::JSON::XS, I would add TO_JSON() to all the classes which does string overloading. I guess my to_array() statement still stand, since only Mojo::Collection has that method.
```
$ ack -l "to_string" lib/ | xargs ack -l '""' | xargs ack -h "^package"
package Mojo::Path;
package Mojo::Date;
package Mojo::Exception;
package Mojo::DOM;
package Mojo::ByteStream;
package Mojo::Parameters;
package Mojo::Cookie;
package Mojo::Home;
package Mojo::URL;
```

**Comment by kraih** [comment]
Ok, after some more discussions on IRC, i don't think this proposal is ready for voting yet. We first need to decide if we want `Mojo::JSON` to behave consistently (or to what degree) with `Cpanel::JSON::XS`.

**Comment by jberger** [comment]
Discussions (at length) on IRC have shown why it is significant to think of this case as distinct from the stringy ones, so on the issue of adding Mojo::Collection::TO_JSON I vote :+1:

**Comment by kraih** [comment]
Thanks, this pull request has been accepted by majority vote.
